students/travis_nelson/lesson05/mailroom_part3.py

```diff
@@ -0,0 +1,163 @@
+                        return response
+            else:
+                raise TypeError
+        except TypeError:
+            print("\nType a name, plz\n")
+
+
+def donation_response(donor):
+    """Returns the donation amount from the user"""
+    keep_asking = True
+    while keep_asking:
+        donation = input("How much did they donate?\n")
+        try:
+            if float(donation):
+                print("\nNice!\n")
+                keep_asking = False
+                return donation
+        except ValueError:
+            print("\nJust a number is fine.\n")
+
+
+def store_new_donation(donor, donation):
+    """Stores the donation amount in the donor database"""
+    updated_donations = donor_db[donor]
+    updated_donations.append(float(donation))
+    donor_db[donor] = updated_donations
+
+
+def sum_donations(donor_name):
+    """Returns sum of all historical donor donations"""
+    return round(sum(donor_db[donor_name]), 2)
+
+
+def sort_key(donor):
+    """Returns the key to sort donations by"""
+    return round(sum(donor_db[donor]))
+
+
+def create_report():
+    """Prints a list of all the donors, their total donation amount,
+    the total number of donations they've made,
+    and the average donation amount,
+    in descending order based on total amount given"""
+    sorted_donor_db = sorted(donor_db, key=sort_key, reverse=True)
+    str_header_formatting = "{:^20}|" * 4
+    str_grid_formatting = "-" * 21*4
+    print(str_header_formatting.format(
+          "Donor Name", "Total Given", "Num Gifts", "Average Gift"))
+    print(str_grid_formatting)
+    for i in sorted_donor_db:
+        donor_name = i
+        donor_total = round(sum(donor_db[i]), 2)
+        number_donations = len(donor_db[i])
+        average_donation = round(donor_total/number_donations)
+        print("{:20} ${:>19} {:>20} ${:>20}".format(
+              donor_name, donor_total, number_donations, average_donation))
+
+
+def generate_all_thanks():
+    """Creates text files containing a thank you letter to each of the donors.
+    Document provides their total amount contributed, as well as their most
+    recent donation amount."""
+    for i in donor_db:
+        most_recent_donation = ''.join(str(e) for e in donor_db[i][-1:])
+        with open(f'{i}.txt', 'w') as f:
+            f.write("Dearest {},\n"
+                    "We are writing to formally thank you for your most recent"
+                    " donation of ${}.\n"
+                    "To date, you have donated ${} to our honorable mission.\n"
+                    "You are truly a valuable patron, and we thank you "
+                    "for your service (and money).\n"
+                    "{}\n"
+                    "{}".format(i,
+                                most_recent_donation,
+                                sum_donations(i),
+                                "Kindest regards,",
+                                "Baron Von Munchausen"))
```

Comment:
the two strings above don't need to be formatted with `format()`, you can use them inline with the string.